imgui-lwjgl3/src/main/java/imgui/glfw/ImGuiImplGlfw.java

```diff
@@ -0,0 +1,427 @@
+            previousKeyCallback = glfwSetKeyCallback(windowId, this::keyCallback);
+            previousCharCallback = glfwSetCharCallback(windowId, this::charCallback);
+        }
+
+        return true;
+    }
+
+    /**
+     * Method to restore {@link org.lwjgl.glfw.GLFW} to it's state prior to calling method {@link ImGuiImplGlfw#init(long, boolean)}.
+     */
+    public void dispose() {
+        if (callbacksInstalled) {
+            glfwSetMouseButtonCallback(windowId, previousMouseButtonCallback);
+            glfwSetScrollCallback(windowId, previousScrollCallback);
+            glfwSetKeyCallback(windowId, previousKeyCallback);
+            glfwSetCharCallback(windowId, previousCharCallback);
+            callbacksInstalled = false;
+        }
+
+        for (int i = 0; i < ImGuiMouseCursor.COUNT; i++) {
+            glfwDestroyCursor(mouseCursors[i]);
+            mouseCursors[i] = 0;
+        }
+    }
+
+    private void updateMousePosAndButtons() {
+        final ImGuiIO io = ImGui.getIO();
+        for (int i = 0; i < ImGuiMouseButton.COUNT; i++) {
+            io.setMouseDown(i, mouseJustPressed[i] || glfwGetMouseButton(windowId, i) != 0);
+            mouseJustPressed[i] = false;
+        }
+
+        final ImVec2 mousePosBackup = new ImVec2();
+        io.getMousePos(mousePosBackup);
+        io.setMousePos(-Float.MAX_VALUE, -Float.MAX_VALUE);
+
+        final boolean focused = glfwGetWindowAttrib(windowId, GLFW_FOCUSED) != 0;
+        if (focused) {
+            if (io.getWantSetMousePos()) {
+                glfwSetCursorPos(windowId, mousePosBackup.x, mousePosBackup.y);
+            } else {
+                try (MemoryStack stack = stackPush()) {
+                    final DoubleBuffer xPos = stack.mallocDouble(1);
+                    final DoubleBuffer yPos = stack.mallocDouble(1);
+                    glfwGetCursorPos(windowId, xPos, yPos);
+                    io.setMousePos((float) xPos.get(), (float) yPos.get());
+                }
+            }
+        }
+    }
+
+    private void updateMouseCursor() {
+        final ImGuiIO io = ImGui.getIO();
+        final boolean noCursorChange = (io.getConfigFlags() & ImGuiConfigFlags.NoMouseCursorChange)
+            == ImGuiConfigFlags.NoMouseCursorChange;
+        final boolean cursorDisabled = glfwGetInputMode(windowId, GLFW_CURSOR) == GLFW_CURSOR_DISABLED;
+        if (noCursorChange || cursorDisabled) {
+            return;
+        }
+
+        final int cursor = ImGui.getMouseCursor();
+        if (cursor == ImGuiMouseCursor.None || io.getMouseDrawCursor()) {
+            glfwSetInputMode(windowId, GLFW_CURSOR, GLFW_CURSOR_HIDDEN);
+        } else {
+            glfwSetCursor(windowId, mouseCursors[cursor] != 0 ? mouseCursors[cursor]
+                : mouseCursors[ImGuiMouseCursor.Arrow]);
+            glfwSetInputMode(windowId, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
+        }
+    }
+
+    private void updateGamepads() {
```

Comment:
You are missing this line https://github.com/ocornut/imgui/blob/master/examples/imgui_impl_glfw.cpp#L311

It fill the `NavInputs` var with zeros.

Reply:
Would that be better done using buffers or with a simple for loop?

e.g. which fails with an `UnsupportedOperationException` on the final line:
```
try (MemoryStack stack = stackPush()) {
            final float[] navInputs = new float[ImGuiNavInput.COUNT];
            io.getNavInputs(navInputs);
            final FloatBuffer buffer = stack.mallocFloat(ImGuiNavInput.COUNT).put(navInputs).flip();
            MemoryUtil.memSet(buffer, 0);
            io.setNavInputs(buffer.array());
        }
```
Verses something like:
```
for (int i = 0; i < ImGuiNavInput.COUNT; i++) {
    io.setNavInputs(i, 0);
}
```
The latter looks cleaner but is it suitably efficient? Still getting to grips with Java -> native memory.

Reply:
> e.g. which fails with an UnsupportedOperationException on the final line:

That's because DirectBytBuffer is mapped to the natively allocated object and can not directly convert it into Java array. The first variant could be rewritten in the next way:
```
You create field with empty array.
private final float[] emptyNavInputsMem = new float[ImGuiNavInput.COUNT]

And use it to fill `NavInputs`.
io.setNavInputs(emptyNavInputsMem);
```

Both variants are acceptable. The second one (with for..in) will be more efficient from the point of the memory footprint and less efficient from the point of runtime performance. Yet that difference is too insignificant and could be ignored. I will be OK with any variant you choose.